src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Basic.Reference.Assemblies;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Features.Workspaces;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.DebugConfiguration;
+using Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+using Microsoft.CodeAnalysis.MetadataAsSource;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.ProjectSystem;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Threading;
+using Microsoft.CodeAnalysis.Workspaces.ProjectSystem;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.Extensions.Logging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+/// <summary>
+/// Workspace which handles file-based programs ('dotnet run app.cs' scenario).
+/// </summary>
+internal sealed class LspFileBasedProgramWorkspace : Workspace, ILspService, ILspWorkspace, ILspFileBasedProgramWorkspace
+{
+    /// <summary>
+    /// A single gate for code that is adding work to <see cref="_projectsToLoadAndReload" />. This is just we don't have code simultaneously trying to load and unload solutions at once.
+    /// </summary>
+    private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
+
+    // TODO: should we use the LanguageServerProjectSystem here instead?
+    private readonly AsyncBatchingWorkQueue<(Uri uri, SourceText documentText, string languageId)> _projectsToLoadAndReload; // TODO: do we need to introduce a shutdown cancellation token for this?
+
+    private readonly ConcurrentDictionary<string, List<LoadedProject>> _loadedProjects = [];
+    private readonly ILspServices _lspServices;
+    private readonly IFileChangeWatcher _fileChangeWatcher;
+    private readonly IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> _dynamicFileInfoProviders;
+    private readonly ProjectTargetFrameworkManager _projectTargetFrameworkManager;
+    private readonly ILoggerFactory _loggerFactory;
+
+    public bool SupportsMutation => true;
+
+    // ILspLogger rationalize
+    ILogger iLogger;
+
+    public LspFileBasedProgramWorkspace(
+        ILspServices lspServices,
+        IFileChangeWatcher fileChangeWatcher,
+        [ImportMany] IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> dynamicFileInfoProviders,
+        ProjectTargetFrameworkManager projectTargetFrameworkManager,
+        ILoggerFactory loggerFactory,
+        IAsynchronousOperationListenerProvider listenerProvider,
+        HostServices hostServices) : base(hostServices, WorkspaceKind.FileBasedPrograms)
```

Comment:
place HostServices at the top.

Reply:
this confused me.  if we have a specific kind for FileBasedProgram workspaces, why do we have code that is stopping the check for misc-worksapces?